Source/WebKit/UIProcess/XR/PlatformXRSystem.cpp

```diff
@@ -132,11 +132,14 @@ void PlatformXRSystem::requestPermissionOnSessionFeatures(const WebCore::Securit
                 protectedThis->m_immersiveSessionMode = mode;
                 protectedThis->m_immersiveSessionGrantedFeatures = grantedFeatures;
                 protectedThis->m_immersiveSessionSecurityOriginData = securityOriginData;
-                protectedThis->setImmersiveSessionState(ImmersiveSessionState::PermissionsGranted);
-            } else
+                protectedThis->setImmersiveSessionState(ImmersiveSessionState::PermissionsGranted, [grantedFeatures = WTFMove(grantedFeatures), completionHandler = WTFMove(completionHandler)](bool) mutable {
+                    completionHandler(WTFMove(grantedFeatures));
```

Comment:
Deleting the completionHandler(...) call on line 139 and adding it to after the async message replies should address the race, as far as I am aware